### PR TITLE
Fixes #30925 - use the organization style Rubocop gem

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Use `theforeman-rubocop` gem to share common RuboCop rules with plugins.
It serves as base for our rubocop styles and rules.

Includes https://github.com/theforeman/foreman/pull/8000. Kudos, @lzap !

Needs:
- [x] move https://github.com/ezr-ondrej/theforeman-rubocop to `theforeman` organization

### CODE REVIEW

**Comment by ares** [comment]
related to https://github.com/theforeman/foreman/pull/7885

File: .rubocop_common.yml
Comment (lzap) [inline .rubocop_common.yml]: If we really want this, I'd suggest to actually split core cops, performance, rails and minitest cops alltogether. Therefore 4 include YAMLs would be provided for optional inclusion.

File: .rubocop_common.yml
Comment (lzap) [inline .rubocop_common.yml]: Also pull those versions out from the include file(s) as plugins can be targeting different versions.

**Comment by ekohl** [comment]
I actually like the idea of a separate gem. Then updating to a newer profile would be be updating the gem dependency, which could be semantically versioned.

Another benefit is that it makes branching easier. Develop is a moving target. That means that if a plugin wants to remain compatible with Foreman 2.0, they should use that. It may be confusing if a plugin's master branch points to a stable branch.

On the Foreman side it would also be a PR that updates the gem version. A PR like https://github.com/theforeman/foreman/pull/8000 would not update `rubocop` to `~> 0.91.0` but rather `foreman-rubocop` to (say) `~> 3.0`.

However, this will likely break hound which doesn't install external gems.

**Comment by ezr-ondrej** [comment]
> However, this will likely break hound which doesn't install external gems.

IMHO with running rubocop as first Github action is better then Hound (more configurable and the response is almost as instant as with Hound).
So I'm fine with the gem.

**Comment by lzap** [comment]
> running rubocop as first Github action is better then Hound

I also think that Hound is something we don't actually leverage too much, it's Rubocop CI job what needs to pass and what we all target for. I am all for ditching Hound and running Rubocop from GHA.

**Comment by lzap** [comment]
> separate gem.

As a plugin author, I hate the idea of someone else bumping dependencies without me knowing and causing my tests to fail forcing me patching many hunks in my codebase because we "agreed" some cop is a good idea. You are still pushing the idea of centralizing this while this is not something that should be centralized. Different people, different codebases, different code-style opinions.

Also a separate gem would make updating a Cop rule huge pain (releasing new gem). We tend to enable Rubocop rules one by one and this would be slow, time consuming and not flexible. :-1: from me

I personally dislike most Style and Layout cops and I will always disable them on my projects. I find Rubocop helpful in other areas like Performance or some Rails cops and few others. In summary, I am not against splitting YAML into smaller YAML files for possible inclusion (I won't make this myself tho). I don't like shipping our rubocop rules or dependencies as a gem - even if we keep YAML files here, bumping Rubocop version *will* break Rubocop tests.

**Comment by ekohl** [comment]
> I also think that Hound is something we don't actually leverage too much, it's Rubocop CI job what needs to pass and what we all target for. I am all for ditching Hound and running Rubocop from GHA.

For what it's worth, that's something @timogoebel has been advocating for a long time. Perhaps it's time for a new vote in a new RFC?



> As a plugin author, I hate the idea of someone else bumping dependencies without me knowing and causing my tests to fail forcing me patching many hunks in my codebase because we "agreed" some cop is a good idea. You are still pushing the idea of centralizing this while this is not something that should be centralized. Different people, different codebases, different code-style opinions.

A plugin is free to choose which gems it depends on. If you run rubocop in a git checkout of the plugin itself, it's just an opt in. It's not that different than the shared config here, with the difference that this shared config doesn't pin the Ruby versions so it's a pain to maintain. A gem could have both a config and the metadata of the versions it's compatible with.

I think that just this shared config isn't very usable for plugins and don't think we should do it this way.

**Comment by ezr-ondrej** [comment]
> You are still pushing the idea of centralizing this while this is not something that should be centralized. Different people, different codebases, different code-style opinions.

I believe that this is where we will disagree. It's still the same project, so it'd be expected to have contributors, that are contributors to the core as well. So it make sense to keep some common style rules.
Although I agree for some projects some rules doesn't make sense. So there should be a way to opt out.

> Also a separate gem would make updating a Cop rule huge pain (releasing new gem). We tend to enable Rubocop rules one by one and this would be slow, time consuming and not flexible.

This is also not entirely true. I believe for such simple gem we can push [automatically](https://github.com/ezr-ondrej/theforeman-rubocop/blob/60b973ab02571a8b738a7f5cae1424237f37260f/.github/workflows/gem-push.yml#L28) AND not necessarily to rubygems, but eg to Github packages, where we can have the luxury to delete versions.


> I personally dislike most Style and Layout cops and I will always disable them on my projects.

As stated above I belive this is wrong mindset, but I agree that the rules should not be hard enforced.


That being said I've tried to draft this gem and publish it at Github packages.
https://github.com/ezr-ondrej/theforeman-rubocop

**Comment by ekohl** [comment]
Indeed. The goal should be to make it easy for plugin authors to follow Foreman's Rubocop rules. Inheriting rules from a gem that you list in your Gemfile / development dependencies is a good way IMHO. You can still list every rule you want and disable it if as a plugin author you disagree with what Foreman uses.

**Comment by lzap** [comment]
For the record, I am not against in ability to share rubocop version/yaml, if this is not enforced and plugin authors can easily pin to specific version so updating Rubocop is under their control. If this allows us to go forward with dropping the ugly hack of enhancing CI rake task with Rubocop for plugins, I am all in.

File: bundler.d/test.rb
Comment (tbrisker) [reply bundler.d/test.rb]: Well, here we have the answer to the question "will hound work with it?"

File: .rubocop_todo.yml
Comment (tbrisker) [inline .rubocop_todo.yml]: Can we continue using `rubocop --auto-gen-config --exclude-limit 0` to auto generate this file? If so, why were these additional cops added manually? ideally the todo file is autogenerated.

File: .github/workflows/rubocop.yml
Comment (ekohl) [inline .github/workflows/rubocop.yml:1]: Looks like you need to install some native libs.

**Comment by ares** [comment]
Given the current situation is worse than this for plugin maintainers, I'm :+1: Plugin maintainers will need to still disable new cops with newer rubocop version, but some may be disabled from core, so it's an improvement.

File: .rubocop.yml
Comment (ekohl) [inline .rubocop.yml:9]: Does it not inherit this from all.yml? Looks like it's falling back to 2.1 now.

**Review by ekohl — COMMENTED** [review]
I do think that eventually we should env up with an almost empty `.rubocop.yml` that only inherits the gem and the todo.

File: .rubocop.yml
Comment (ezr-ondrej) [reply .rubocop.yml:9]: It should, I think it was just picking up on a wrong gem version.

File: .github/workflows/rubocop.yml
Comment (ezr-ondrej) [reply .github/workflows/rubocop.yml:1]: Yeah, I didn't disable the groups properly. I wanted to end-up with as minimal gem set as possible.

File: bundler.d/test.rb
Comment (ezr-ondrej) [reply bundler.d/test.rb]: Yeah, let's just get rid of Hound? :thinking:

**Comment by ezr-ondrej** [comment]
> I do think that eventually we should env up with an almost empty `.rubocop.yml` that only inherits the gem and the todo.

Agreed, but that involve a disscusion around what cops should be really disabled.
One way would be to just enable them and add them to todo.

File: .github/workflows/rubocop.yml
Comment (lzap) [inline .github/workflows/rubocop.yml]: Is there a particular reason why you don't use the `actions/cache@` step? It makes everything much faster.

File: bundler.d/test.rb
Comment (lzap) [inline bundler.d/test.rb]: Please add the theforeman account to the:

https://rubygems.org/gems/theforeman-rubocop

This one:

https://rubygems.org/profiles/theforeman

Also I encourage you to start with 1.0.0 but this is just my opinion :-)

File: .rubocop.yml
Comment (lzap) [inline .rubocop.yml]: Why are you leaving some rules here? What is the key to put it here and there? Can you document this key?

File: bundler.d/test.rb
Comment (lzap) [reply bundler.d/test.rb]: Also link to homepage does not work (404): https://github.com/theforeman/theforeman-rubocop

Also the description could be little bit better:

> This gem guides and shares rubocop style configs.

Something like "Shared Rubocop configuration for theforeman.org family of projects" or something like that.

Comment (ezr-ondrej) [reply bundler.d/test.rb]: I've added theforeman to gem owners, moved the repo to the theforeman org (so the homepage works now) and I will update the description :)

File: .rubocop.yml
Comment (ezr-ondrej) [reply .rubocop.yml]: TL;DR: I'm not sure what to move to `.rubocop_todo.yml` and what to the gem.

I've moved only those, that I remember Foreman maintainers having disscussion around, others I've left in foreman and I'd like others to tell me if to move them or not. My idea currently is to create issues for them in the gem. But I'll go over them and try gather more info on them prior to that.

Good example is a hash rockets:
It is disabled here, but only because [we've decided](https://community.theforeman.org/t/hash-rocket-syntax/5865) against `rubocop -a` on that rule, but we would like to slowly get there, so the common config should prefer `key: a` style over rockets, so it should move to `.rubocop_todo.yml`.

File: bundler.d/test.rb
Comment (ekohl) [reply bundler.d/test.rb]: I think semver suggests starting with `0.1.0` so you can easily push `0.1.1` without everyone having to update their gemfiles.

Comment (ezr-ondrej) [reply bundler.d/test.rb]: I'd like the version `0.1.0` be clean core `.rubocop.yml`.
That is by no means work done as I belive there is many rubocop styles we want to enable/configure, but it would be a good start IMHO.
But yeah, before merging this PR we should release `0.1.0`.

Comment (lzap) [reply bundler.d/test.rb]: > I've added theforeman to gem owners

Still don't see these at https://rubygems.org/gems/theforeman-rubocop please do this we can't merge until we have full control of the rubygem. It bites!

**Comment by lzap** [comment]
Ok where are we at? I'd like to push this forward. I think the poll showed that people are okay with the current name.

The division into all/default group is probably okay, I'd rather see something like loose vs strict but it's fine. Plugin authors can override themselves.

**Comment by ezr-ondrej** [comment]
> Ok where are we at? I'd like to push this forward. I think the poll showed that people are okay with the current name.

I'd say let's rename the cops, release gem and lets get this in, cops move to the gem can happen over time? :shrug:

File: .github/workflows/rubocop.yml
Comment (ezr-ondrej) [reply .github/workflows/rubocop.yml]: I've used `bundler-cache`, that we've discovered recently in https://github.com/theforeman/foreman_ansible/pull/368 and works better then manual usage of `actions/cache@`

**Comment by ezr-ondrej** [comment]
We are :green_apple:

**Comment by ezr-ondrej** [comment]
[test katello] doesn't look like RuboCop failure, but to be sure #IdontlikeuKatello

**Comment by ezr-ondrej** [comment]
[test katello]

File: .github/workflows/rubocop.yml
Comment (ekohl) [inline .github/workflows/rubocop.yml]: I would consider the [Github Actions formatter](https://docs.rubocop.org/rubocop/1.4/formatters.html#github-actions-formatter). Haven't tried it myself and perhaps you need a newer Rubocop version than we use now.

File: .github/workflows/rubocop.yml
Comment (ekohl) [inline .github/workflows/rubocop.yml]: 2.6 is an odd version to test with. Currently all production runs on 2.5 but the next target will be 2.7. I don't think we'll ever ship with 2.6.
```suggestion
          ruby-version: 2.7
```

File: bundler.d/test.rb
Comment (ekohl) [inline bundler.d/test.rb]: From the packaging side https://github.com/theforeman/foreman-packaging/pull/6012 should be merged to deal with the different way of group handling. Note that I wrote that even before I knew it was needed here.

**Review by ekohl — COMMENTED** [review]
Overall it looks good, some small suggestions inline.

File: bundler.d/test.rb
Comment (ezr-ondrej) [reply bundler.d/test.rb]: oh, right, well we don't need this group for now, so let's just leave it at test :)
We will just bundle few more gems for now. I'll try to remember to update it once the packaging is there (or I'll open followup right away)

File: .github/workflows/rubocop.yml
Comment (ezr-ondrej) [reply .github/workflows/rubocop.yml]: Ah, thanks for the info, I'll go for 2.5 then, as that's the TargetVersion for the Cops.

File: .github/workflows/rubocop.yml
Comment (ezr-ondrej) [reply .github/workflows/rubocop.yml]: Yeah, that is definitelly out of reach now, but I'm all in for upgrading the rubocop once we get rid of hound :+1:

**Review by lzap — APPROVED** [review]
Let's proceed with this one. Merging, thanks.

**Pull request merged**
